common/lib/capa/capa/responsetypes.py

```diff
@@ -1362,6 +1572,79 @@ def check_string_backward(self, expected, given):
+        The forms show below match in the order given, and the first matching one stops the matching.
+        <stringresponse answer="A" type="ci">
+          <correcthint>hint1</correcthint>                         <!-- hint for correct answer -->
+          <additional_answer answer="B">hint2</additional_answer>  <!-- additional_answer with its own hint -->
+          <stringequalhint answer="C">hint3</stringequalhint>      <!-- string matcher/hint for an incorrect answer -->
+          <regexphint answer="FG+">hint4</regexphint>              <!-- regex matcher/hint for an incorrect answer -->
+          <textline size="20"/>
+        </stringresponse>
+        The "ci" and "regexp" options are inherited from the parent stringresponse as appropriate.
+        """
+        if self.answer_id in student_answers:
+            student_answer = student_answers[self.answer_id]
+            # Note the atypical case of using self.id instead of self.answer_id
+            responses = self.xml.xpath('//stringresponse[@id=$id]', id=self.id)
+            if responses:
+                response = responses[0]
+
+                # First call the existing check_string to see if this is a right answer by that test.
+                # It handles the various "ci" "regexp" cases internally.
+                expected = response.get('answer').strip()
+                if self.check_string([expected], student_answer):
+                    hint_node = response.find('./correcthint')
+                    if hint_node is not None:
+                        new_cmap[self.answer_id]['msg'] += self.make_hint_div(hint_node, True)
+                    return
+
+                # Then look for additional answer with an answer= attribute
+                for node in response.findall('./additional_answer'):
+                    if self.match_hint_node(node, student_answer, self.regexp, self.case_insensitive):
+                        correct_hint = node.find('./correcthint')
+                        new_cmap[self.answer_id]['msg'] += self.make_hint_div(correct_hint, True)
+                        return
+
+                # stringequalhint and regexphint represent wrong answers
+                for node in response.findall('./stringequalhint'):
+                    if self.match_hint_node(node, student_answer, False, self.case_insensitive):
+                        new_cmap[self.answer_id]['msg'] += self.make_hint_div(node, False)
+                        return
+
+                for node in response.findall('./regexphint'):
+                    if self.match_hint_node(node, student_answer, True, self.case_insensitive):
+                        new_cmap[self.answer_id]['msg'] += self.make_hint_div(node, False)
+                        return
+
+    def match_hint_node(self, node, given, regex_mode, ci_mode):
+        """
+        Given an xml node such as additional_answer or regexphint, which contain an answer=,
+        returns True if the given student answer is a match.
```

Comment:
Docstring is a little light. Maybe include descriptions of what each parameter represents? Would make it easier to read the code, because i'm a bit shaky on what "regex_mode" and "ci_mode" mean